projects/origin-release/release-image-builder/build.sh

```diff
@@ -0,0 +1,60 @@
+#!/bin/bash -xeu
+
+env | sort
+
+if [ ! -e "${DOCKER_SOCKET}" ]; then
+  echo "Docker socket missing at ${DOCKER_SOCKET}"
+  exit 1
+fi
+
+arch=${ARCH:-x86_64}
+namespace=$(echo "${BUILD}" | jq --raw-output '.metadata.namespace')
+base_image="docker-registry.default.svc:5000/${namespace}/${BASE_IMAGE_STREAM_TAG}"
+tag=$(echo "${BUILD}" | jq --raw-output '.status.outputDockerImageReference')
+host_arch=$(arch)
+build_dir=$(mktemp --directory)
+
+git clone --recursive "${SOURCE_REPOSITORY}" "${build_dir}"
+if [ $? != 0 ]; then
+  echo "Error trying to fetch git source: ${SOURCE_REPOSITORY}"
+  exit 1
+fi
+
+pushd "${build_dir}"
+
+if [ -n "${SOURCE_REF}" ]; then
+  git checkout "${SOURCE_REF}"
+  if [ $? != 0 ]; then
+    echo "Error trying to checkout branch: ${SOURCE_REF}"
+    exit 1
+  fi
+fi
+
+if [[ -n "${SOURCE_CONTEXT_DIR}" ]]; then
+  pushd "${SOURCE_CONTEXT_DIR}"
+fi
+
+if [[ -n "${PULL_DOCKERCFG_PATH}" ]]; then
+  cp "${PULL_DOCKERCFG_PATH}/.dockercfg" "${HOME}/.dockercfg"
+fi
+
+docker pull "${base_image}"
+
+if [[ "${host_arch}" != "${arch}" ]]; then
+  imagebuilder --mount /usr/bin/qemu-${arch}-static:/usr/bin/qemu-${arch}-static --from "${base_image}" -t "${tag}" .
+else
+  imagebuilder --from "${base_image}" -t "${tag}" .
+fi
+
+if [[ -n ${COPY_DEST} ]]; then
+  if [[ -e "/build/copy-dest-creds/.dockercfg" ]]; then
+    cp /build/copy-dest-creds/.dockercfg "${HOME}/.dockercfg"
+  fi
+  skopeo copy docker-daemon:${tag} docker://${COPY_DEST}
+fi
+
+if [[ -n "${PUSH_DOCKERCFG_PATH}" ]]; then
+  cp "${PUSH_DOCKERCFG_PATH}/.dockercfg" "${HOME}/.dockercfg"
+fi
+
+docker push "${tag}"
```

Comment:
Originally, I was attempting to just push an image to dockerhub and use a scheduled imagestream tag to trigger later builds, but I hit two issues with this. The first is that `docker push` from within a custom build image seems to be broken, and the second was a very noticeable delay before later builds are triggered, since they are only triggered after the scheduled IS update is performed. Using skopeo to copy the image to docker hub solved the `docker push` issue, but I added a second push to the image stream to make the delay less noticeable. Unfortunately, this also has the side-effect of making the builds take longer as it has to push to both registries.

If there are no issues with having a delay between pushing the base image, or arch specific images and triggering the subsequent builds, I can revert to the single push and scheduled IS updates.

Reply:
I also updated the base image to be pushed to docker hub, to allow for more easily allowing developers to build their own "release" image outside of the CI env.

Reply:
I'm not sure pushing from within a build makes any sense. Why does the delay matter? Is this a critical path for something? Also shouldn't these intermediate images _not_ be pushed to any registry other than the cluster internal one?

Reply:
> I'm not sure pushing from within a build makes any sense.

If you mean pushing to both registries, then I'm likely to agree. Custom Builds require that at least one push happens from within the build, though (well, if there is an output defined for the build).

> Why does the delay matter? Is this a critical path for something?

I don't think it's critical path for anything, but it did help with test iteration and decreases the overall time it takes to publish a new release image on request. If the delays for those use cases aren't overly important, then I'm happy to remove.

> Also shouldn't these intermediate images not be pushed to any registry other than the cluster internal one?

If we do not publish the base images to dockerhub, then the release image is not able to be built locally using docker build without first building the base image and tagging it appropriately. If we aren't concerned with being able to build the images easily outside of the CI env, then I'm good with updating to only publish to the internal registry.

Reply:
I'm not sure we care a huge amount about people recreating these envs locally -- but @smarterclayton should weigh in